NEW Add additional session handler implementations

GuySartorelli opened the pull request:
Doing this in several commits - DO NOT SQUASH

Commit 1: The cache-based session handler that can be used for Redis and Memcached
Commit 2: The database session handler
Commit 3: Add new environment variable to set the save handler
Commit 4: Injector allows backticks in the `factory` key now. This isn't a breaking change, because using backticks for that before would be trying to get a class name that has backticks in it, which PHP doesn't allow.


## Issue

- https://github.com/silverstripe/silverstripe-framework/issues/11719

GuySartorelli committed:
NEW Add cache-based session handler implementation

The intention is to use in-memory cache adapters such as Redis and
Memcached - though theoretically any PSR16 adapter would work.

GuySartorelli commented on src/Control/SessionHandler/CacheSessionHandler.php:
Implementing the session handler for cache adapters as a single abstracted handler ended up being simpler than if I tried to implement Redis and Memcached handlers separately - with the bonus that it's also easier to test since we can use Symfony's `ArrayAdapter`.

GuySartorelli commented on src/Control/SessionHandler/AbstractSessionHandler.php:
The session lifetime is the same across all session handlers so it made sense to move into its own abstract class (pulled from `FileSessionHandler`).

GuySartorelli commented on tests/php/ORM/DataListTest.php:
```diff
@@ -865,9 +865,7 @@ public static function provideDefaultSort(): array
-    /**
-     * @dataProvider provideDefaultSort
-     */
+    #[DataProvider('provideDefaultSort')]
```
Unrelated MNT fix. Noticed a warning about this in the test run so I figured I may as well just get it fixed.

GuySartorelli commented on tests/php/Control/SessionHandler/FileSessionHandlerTest.php:
```diff
@@ -481,7 +481,7 @@ public function testGc(int $gcLifetime, int $configLifetime, array $sessionFiles
 
         ini_set('session.gc_maxlifetime', $gcLifetime);
-        Session::config()->set('lifetime', $configLifetime);
+        Session::config()->set('timeout', $configLifetime);
```
This and the implementation in `FileSessionHandler` were looking at the wrong config name.

GuySartorelli committed:
NEW Add database session handler implementation

This is an easy option for multi-server hosting environments, though is
likely less performant than the other session handlers on offer.

GuySartorelli commented on src/Control/SessionHandler/DatabaseSessionHandler.php:
```diff
@@ -0,0 +1,222 @@
+        return $result === true || $result === 1 || $result === '1';
+    }
+
+    private function isDatabaseReady()
+    {
+        if (!DB::connection_attempted() || !DB::is_active()) {
+            return false;
+        }
+        return DB::get_schema()->hasTable(static::config()->get('table_name'));
+    }
```
This was originally going to use the new `DataObjectSchema::tablesAreReadyForClass()` method, but since we aren't using a `DataObject` (see phpdoc comment for `requireTable()` for reasons why not) that's not suitable.
Ultimately since the schema for this table isn't configurable, the table will either exist and have the correct schema, or not exist.

GuySartorelli committed:
ENH Add environment variable to set session save handler.

This allows hosting providers to set specific session save handlers if
they have infrastructure set up for one that is preferred (e.g. if using
redis).

GuySartorelli commented on tests/php/Control/SessionTest.php:
```diff
@@ -16,7 +16,11 @@
 use SilverStripe\Core\Config\Config;
 use SilverStripe\Core\Injector\Injector;
 use PHPUnit\Framework\Attributes\DataProvider;
+use PHPUnit\Framework\Attributes\PreserveGlobalState;
```
Noticed this `use` statement was missing... I have no idea why it wasn't causing CI failures but this class is referenced in some of the tests higher up in this class.

emteknetnz commented on src/Control/SessionHandler/CacheSessionHandler.php:
```diff
@@ -0,0 +1,87 @@
+ */
+class CacheSessionHandler extends AbstractSessionHandler
+{
+    private CacheInterface $cacheHandler;
```
```suggestion
    private CacheInterface $cache;
```

Note sure what the 'Handler' suffix is for? Also update all the references to it

GuySartorelli replied:
Oops should have been `$cacheAdapter` which is what I'll set it to.

emteknetnz replied:
Where are you getting the word "adapter" from? That word is not present on https://www.php-fig.org/psr/psr-16/

`CacheInterface` implies that it's an interface for a cache?

I *think* there is a concept of an "adapter" if you want to adapt a PSR-6 cache interface to a PSR-16 cache interface, though a PSR-16 compatible cache can also just be a stand-alone cache with no compatibility to a PSR-6 cache.

ie. the PSR-16 compatible object being passed in could be a cache wrapped in an adapter, or just a cache, so $cache would be a better name as $cacheAdapter could be wrong.

GuySartorelli replied:
I don't think the name matters that much :sweat_smile: we both know what this variable holds. If you really really really want me to change it to `$cache` I will but I think we're just spinning wheels at this point.

> Where are you getting the word "adapter" from?

It's common parlance when dealing with symfony cache *stuff*, which is the primary cache *stuff* we tend to use.

"cache" on its own is an overloaded word. I tend to think of "cache" as the content that is _stored_ or the place _where_ it's stored.... this variable represents neither of those. It represents the API that is used to take some data and store it in the cache. Adapter seems like a good word for that. I guess `$cacheTransport` would also make sense for that but I've not seen that terminology used anywhere before.

emteknetnz replied:
Change it to $cache please :-)

GuySartorelli replied:
Sure. Done.

emteknetnz commented on _config/session.yml:
```diff
@@ -0,0 +1,39 @@
+SilverStripe\Core\Injector\Injector:
+  SilverStripe\Control\SessionHandler\CacheSessionHandler:
+    constructor:
+      cacheAdapter: '%$Psr\SimpleCache\CacheInterface.session-handler'
```
```suggestion
      cache: '%$Psr\SimpleCache\CacheInterface.session-handler'
```
This didn't actually match "cacheHandler", which I've recommended be changed to "cache"

emteknetnz replied:
Still needs to be done

GuySartorelli replied:
Oops sorry, forgot about this one.
Done now.

emteknetnz commented on src/Control/SessionHandler/DatabaseSessionHandler.php:
```diff
@@ -0,0 +1,218 @@
+    public function read(#[SensitiveParameter] string $id): string|false
+    {
+        if (!$this->isDatabaseReady()) {
+            return false;
```
I'm wondering if we'd be better off just throwing an exception if isDatabaseRead() returns false. What happens in all these methods like open() / gc() / read() etc if `false` is returned?  Is there some risk of things getting out of sync and acting weird if we don't have a hard failure?

GuySartorelli replied:
The API for this method and other methods that can return false is defined by PHP itself - if the intention was for hard exceptions to be thrown, PHP would have added a `@throws` to the PHPdoc.

I'll add some logging though, that seems sensible.
The new logging is similar to logging in `FileSessionHandler`.

emteknetnz commented on src/Control/SessionHandler/CacheSessionHandler.php:
```diff
@@ -0,0 +1,87 @@
+use Psr\SimpleCache\CacheInterface;
+
+/**
+ * Session save handler that stores session data in an in a PSR-16 cache adapter.
```
```suggestion
 * Session save handler that stores session data in an in a PSR-16 cache.
```

GuySartorelli replied:
Done

emteknetnz commented on tests/php/Control/SessionHandler/CacheSessionHandlerTest.php:
```diff
@@ -0,0 +1,142 @@
+    #[DataProvider('provideRead')]
+    public function testRead(string $sessionID, string $expected): void
+    {
+        $cacheAdapter = new Psr16Cache(new ArrayAdapter());
+        $handler = new CacheSessionHandler($cacheAdapter);
+        $cacheAdapter->set('existing-session', 'some-data');
```
```suggestion
        $cache = new Psr16Cache(new ArrayAdapter());
        $handler = new CacheSessionHandler($cache);
        $cache->set('existing-session', 'some-data');
```

GuySartorelli replied:
That is literally a cache adapter in this case.... but done anyway to avoid ping pong

emteknetnz commented on tests/php/Control/SessionHandler/CacheSessionHandlerTest.php:
```diff
@@ -0,0 +1,142 @@
+
+    public function testReadExpired(): void
+    {
+        $cacheAdapter = new Psr16Cache(new ArrayAdapter());
+        $handler = new CacheSessionHandler($cacheAdapter);
+        $cacheAdapter->set('existing-session', 'some-data', -1);
+        $this->assertSame('', $handler->read('existing-session'));
```
```suggestion
        $cache = new Psr16Cache(new ArrayAdapter());
        $handler = new CacheSessionHandler($cache);
        $cache->set('existing-session', 'some-data', -1);
        $this->assertSame('', $handler->read('existing-session'));
```

GuySartorelli replied:
Done

emteknetnz commented on tests/php/Control/SessionHandler/CacheSessionHandlerTest.php:
```diff
@@ -0,0 +1,142 @@
+    #[DataProvider('provideWrite')]
+    public function testWrite(string $sessionID): void
+    {
+        $cacheAdapter = new Psr16Cache(new ArrayAdapter());
+        $handler = new CacheSessionHandler($cacheAdapter);
+        $cacheAdapter->set('existing-session', 'some-data');
+        $handler->write($sessionID, 'updated-data');
+        $this->assertSame('updated-data', $cacheAdapter->get($sessionID));
```
```suggestion
        $cache = new Psr16Cache(new ArrayAdapter());
        $handler = new CacheSessionHandler($cache);
        $cache->set('existing-session', 'some-data');
        $handler->write($sessionID, 'updated-data');
        $this->assertSame('updated-data', $cache->get($sessionID));
```

GuySartorelli replied:
Done

emteknetnz commented on tests/php/Control/SessionHandler/CacheSessionHandlerTest.php:
```diff
@@ -0,0 +1,142 @@
+    #[DataProvider('provideDestroy')]
+    public function testDestroy(string $sessionID): void
+    {
+        $cacheAdapter = new Psr16Cache(new ArrayAdapter());
+        $handler = new CacheSessionHandler($cacheAdapter);
+        $cacheAdapter->set('existing-session', 'some-data');
+        $this->assertTrue($handler->destroy($sessionID));
+        $this->assertNull($cacheAdapter->get($sessionID));
```
```suggestion
        $cache = new Psr16Cache(new ArrayAdapter());
        $handler = new CacheSessionHandler($cache);
        $cache->set('existing-session', 'some-data');
        $this->assertTrue($handler->destroy($sessionID));
        $this->assertNull($cache->get($sessionID));
```

GuySartorelli replied:
Done

emteknetnz commented on tests/php/Control/SessionHandler/CacheSessionHandlerTest.php:
```diff
@@ -0,0 +1,142 @@
+
+    public function testUpdateTimestamp(): void
+    {
+        $arrayAdapter = new ArrayAdapter();
+        $cacheAdapter = new Psr16Cache($arrayAdapter);
+        $handler = new CacheSessionHandler($cacheAdapter);
+        $cacheAdapter->set('existing-session', 'some-data', 999999999);
+
+        $this->assertTrue($handler->updateTimestamp('existing-session', 'new content'));
+        $this->assertTrue($cacheAdapter->has('existing-session'));
```
```suggestion
        $arrayAdapter = new ArrayAdapter();
        $cache = new Psr16Cache($arrayAdapter);
        $handler = new CacheSessionHandler($cache);
        $cache->set('existing-session', 'some-data', 999999999);

        $this->assertTrue($handler->updateTimestamp('existing-session', 'new content'));
        $this->assertTrue($cache->has('existing-session'));
```

GuySartorelli replied:
Done

emteknetnz commented on tests/php/Control/SessionHandler/CacheSessionHandlerTest.php:
```diff
@@ -0,0 +1,142 @@
+        // and Symfony's cache stuff doesn't use our internal DateTime so we can't set a mock now.
+        $this->assertLessThan(microtime(true) + 999999, $expiry);
+        $this->assertGreaterThan(microtime(true), $expiry);
+        $this->assertSame('new content', $cacheAdapter->get('existing-session'));
```
```suggestion
        $this->assertSame('new content', $cache->get('existing-session'));
```

GuySartorelli replied:
Done

emteknetnz commented on tests/php/Control/SessionHandler/DatabaseSessionHandlerTest.php:
```diff
@@ -0,0 +1,275 @@
+    {
+        // Add the sessions table
+        $handler = new DatabaseSessionHandler();
+        DB::get_schema()->schemaUpdate(fn () => $handler->requireTable());
```
```suggestion
        DB::get_schema()->schemaUpdate(fn() => $handler->requireTable());
```

GuySartorelli replied:
Done - if the linter complains I'll undo it

emteknetnz commented on tests/php/Control/SessionHandler/DatabaseSessionHandlerTest.php:
```diff
@@ -0,0 +1,275 @@
+    {
+        parent::setUp();
+        $expiry = DBDatetime::now()->getTimestamp() + 1000;
+        DB::query('UPDATE "_sessions" SET "Expiry" = ' . $expiry . ' WHERE "ID" = \'valid\'');
```
```suggestion
        $tableName = DatabaseSessionHandler::config()->get('table_name');
        DB::query('UPDATE "' . $tableName . '" SET "Expiry" = ' . $expiry . ' WHERE "ID" = \'valid\'');
```

GuySartorelli replied:
Done though I don't fully agree with this one.

emteknetnz commented on tests/php/Control/SessionHandler/DatabaseSessionHandlerTest.php:
```diff
@@ -0,0 +1,275 @@
+
+    public function testIdIsPrimaryKey(): void
+    {
+        $result = DB::query('SHOW KEYS FROM "_sessions" WHERE "Key_name" = \'PRIMARY\'');
```
```suggestion
        $tableName = DatabaseSessionHandler::config()->get('table_name');
        $result = DB::query('SHOW KEYS FROM "' . $tableName . '" WHERE "Key_name" = \'PRIMARY\'');
```

GuySartorelli replied:
Done

emteknetnz commented on tests/php/Control/SessionHandler/DatabaseSessionHandlerTest.php:
```diff
@@ -0,0 +1,275 @@
+
+        if ($sessionID === 'new-session') {
+            // Make sure no new file is created for new sessions
+            $result = DB::query('SELECT "ID" FROM "_sessions" WHERE "ID" = \'new-session\'');
```
```suggestion
            $tableName = DatabaseSessionHandler::config()->get('table_name');
            $result = DB::query('SELECT "ID" FROM "' . $tableName .'" WHERE "ID" = \'new-session\'');
```

GuySartorelli replied:
Done

emteknetnz commented on tests/php/Control/SessionHandler/DatabaseSessionHandlerTest.php:
```diff
@@ -0,0 +1,275 @@
+        $expiry = $now->getTimestamp() + $expectedLifetime;
+
+        $this->assertTrue($handler->write($sessionID, 'New content now'));
+        $result = DB::query('SELECT * FROM "_sessions" WHERE "ID" = \'' . $sessionID . '\'');
```
```suggestion
        $tableName = DatabaseSessionHandler::config()->get('table_name');
        $result = DB::query('SELECT * FROM ". $tableName .'" WHERE "ID" = \'' . $sessionID . '\'');
```

GuySartorelli replied:
Done

emteknetnz commented on tests/php/Control/SessionHandler/DatabaseSessionHandlerTest.php:
```diff
@@ -0,0 +1,275 @@
+        $handler = new DatabaseSessionHandler();
+
+        $this->assertTrue($handler->destroy($sessionID));
+        $result = DB::query('SELECT "ID" FROM "_sessions" WHERE "ID" = \'' . $sessionID . '\'');
```
```suggestion
        $tableName = DatabaseSessionHandler::config()->get('table_name');
        $result = DB::query('SELECT "ID" FROM "' . $tableName . '" WHERE "ID" = \'' . $sessionID . '\'');
```

GuySartorelli replied:
Done

emteknetnz commented on tests/php/Control/SessionHandler/DatabaseSessionHandlerTest.php:
```diff
@@ -0,0 +1,275 @@
+        $numDeleted = $handler->gc(1);
+        // Check it deleted the right things
+        foreach ($expectDeleted as $sessionID) {
+            $result = DB::query('SELECT "ID" FROM "_sessions" WHERE "ID" = \'' . $sessionID . '\'');
```
```suggestion
            $result = DB::query('SELECT "ID" FROM "'  .$tableName . '" WHERE "ID" = \'' . $sessionID . '\'');
```

GuySartorelli replied:
Done

emteknetnz commented on tests/php/Control/SessionHandler/DatabaseSessionHandlerTest.php:
```diff
@@ -0,0 +1,275 @@
+        DBDatetime::set_mock_now($now);
+        // $lifeToDate represents how long ago the session was last written to
+        // so we need to calculate its expiry date based on how much lifetime is left.
+        foreach ($sessionLifetimeMap as $sessionID => $lifeToDate) {
+            $expiry = $now->getTimestamp() + ($lifetime - $lifeToDate);
+            DB::query(sprintf(
+                'INSERT INTO "_sessions" ("ID", "Data", "Expiry") VALUES (\'%s\', \'%s\', \'%s\')',
```
```suggestion
        $tableName = DatabaseSessionHandler::config()->get('table_name');
        foreach ($sessionLifetimeMap as $sessionID => $lifeToDate) {
            $expiry = $now->getTimestamp() + ($lifetime - $lifeToDate);
            DB::query(sprintf(
                'INSERT INTO "' . $tableName . '" ("ID", "Data", "Expiry") VALUES (\'%s\', \'%s\', \'%s\')',
```

GuySartorelli replied:
Done (hard to see what's changed in the github diff but I assume it's just the table name)

emteknetnz commented on tests/php/Control/SessionHandler/DatabaseSessionHandlerTest.php:
```diff
@@ -0,0 +1,275 @@
+            $this->assertTrue($handler->updateTimestamp($sessionID, 'new content'));
+        });
+
+        $result = DB::query('SELECT * FROM "_sessions" WHERE "ID" = \'' . $sessionID . '\'');
```
```suggestion
        $tableName = DatabaseSessionHandler::config()->get('table_name');
        $result = DB::query('SELECT * FROM "' . $tableName . '" WHERE "ID" = \'' . $sessionID . '\'');
```

GuySartorelli replied:
Done

emteknetnz commented on src/Control/SessionHandler/DbBuildSessionExtension.php:
```diff
@@ -0,0 +1,66 @@
+    /**
+     * This extension hook is on TestSessionEnvironment, which is used by behat but not by phpunit.
+     * For whatever reason, behat doesn't build the db the normal way, so we can't rely on the below
+     * onAfterbuild being run in that scenario.
```
```suggestion
     * onAfterBuild being run in that scenario.
```

GuySartorelli replied:
Done

emteknetnz commented on src/Control/SessionHandler/DbBuildSessionExtension.php:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    /**
+     * This extension hook is on DbBuild, after building the database.
```
```suggestion
     * This extension hook is on DbBuild::doBuild(), after building the database.
```

GuySartorelli replied:
Done

GuySartorelli replied:
Done

emteknetnz commented on src/Control/SessionHandler/DbBuildSessionExtension.php:
```diff
@@ -0,0 +1,66 @@
+     * For whatever reason, behat doesn't build the db the normal way, so we can't rely on the below
+     * onAfterbuild being run in that scenario.
+     */
+    protected function onAfterStartTestSession()
```
```suggestion
    protected function onAfterStartTestSession(): void
```

GuySartorelli replied:
Done

emteknetnz commented on src/Control/SessionHandler/DatabaseSessionHandler.php:
```diff
@@ -0,0 +1,230 @@
+        return $result === true || $result === 1 || $result === '1';
+    }
+
+    private function isDatabaseReady()
```
```suggestion
    private function isDatabaseReady(): bool
```

GuySartorelli replied:
Done

emteknetnz commented on src/Control/SessionHandler/DatabaseSessionHandler.php:
```diff
@@ -0,0 +1,230 @@
+     * 3. We don't want sessions to be used in any other ways that DataObjects are often
+     * 4. We only want to build the table if this is the configured save handler
+     */
+    public function requireTable()
```
```suggestion
    public function requireTable(): void
```
Covariance allows more specific return types

GuySartorelli replied:
Done

emteknetnz commented on _config/session.yml:
```diff
@@ -0,0 +1,39 @@
+    constructor:
+      cacheAdapter: '%$Psr\SimpleCache\CacheInterface.session-handler'
+
+  SilverStripe\Core\Cache\RedisCacheFactory:
+    constructor:
+      logger: '%$Psr\Log\LoggerInterface'
+
+  SilverStripe\Core\Cache\MemcachedCacheFactory:
+    constructor:
+      logger: '%$Psr\Log\LoggerInterface'
```
Seems like these two should be in a separate (new?) config file, they're not specific to sessions i.e. could be used for other caches?

GuySartorelli replied:
These are specifically here for sessions. I suppose they could be used for other things but they're not intended to be used for other things.

emteknetnz commented on tests/php/Control/SessionHandler/DatabaseSessionHandlerTest.php:
```diff
@@ -0,0 +1,282 @@
+    public function testGc(int $gcLifetime, int $configLifetime, array $sessionLifetimeMap, array $expectDeleted): void
+    {
+        $tableName = DatabaseSessionHandler::config()->get('table_name');
+        ini_set('session.gc_maxlifetime', $gcLifetime);
```
Shouldn't this be reset after the test?

GuySartorelli replied:
I don't think it particularly matters (sessions aren't used in CI anyway and this doesn't actually set anything permanently, only for the current process) - but have reset anyway for this and `FileSessionHandlerTest` to avoid further ping pong.

emteknetnz commented on src/Control/SessionHandler/DbBuildSessionExtension.php:
```diff
@@ -0,0 +1,66 @@
+
+    private function getSessionHandler(): ?DatabaseSessionHandler
+    {
+        $sessionHandlerServiceName = Session::getSaveHandler();
```
Session::getSaveHandler() returns an object, not a service name?

Probably means we don't need the  `$sessionHandler = Injector::inst()->get($sessionHandlerServiceName);` below?

GuySartorelli replied:
Oops, I must have had an earlier implementation of `getSaveHandler()` and not updated this or something. Fixed.

GuySartorelli commented:
The CI failures are unrelated - compare with [the 6 branch](https://github.com/silverstripe/silverstripe-framework/actions/runs/16483361801/job/46602913773)

emteknetnz reviewed (changes requested):
I set my .env to 

`SS_SESSION_SAVE_HANDLER_CLASS="SilverStripe\Control\SessionHandler\DatabaseSessionHandler"`

and I ran dev/build flush

I got this on the login form when I went to /admin (which then redirects to /Security/login?BackURL=%2Fadmin%2Fpages)

```text
         // Session start emits a cookie, but only if there's no existing session. If there is a session timeout
336                 // tied to this request, make sure the session is held for the entire timeout by refreshing the cookie age.
337                 if ($cookieParams['lifetime'] && $this->requestContainsSessionId()) {
338                     Cookie::set(
339                         session_name(),

Trace

    session_start()
    Session.php:333
    SilverStripe\Control\Session->start(SilverStripe\Control\HTTPRequest)
    Session.php:252
    SilverStripe\Control\Session->init(SilverStripe\Control\HTTPRequest)
    SessionMiddleware.php:17
    SilverStripe\Control\Middleware\SessionMiddleware->process(SilverStripe\Control\HTTPRequest, Closure)
    HTTPMiddlewareAware.php:62
    SilverStripe\Control\Director->SilverStripe\Control\Middleware\{closure}(SilverStripe\Control\HTTPRequest)
    AllowedHostsMiddleware.php:71
    SilverStripe\Control\Middleware\AllowedHostsMiddleware->process(SilverStripe\Control\HTTPRequest, Closure)
    HTTPMiddlewareAware.php:62
    SilverStripe\Control\Director->SilverStripe\Control\Middleware\{closure}(SilverStripe\Control\HTTPRequest)
    TrustedProxyMiddleware.php:179
    SilverStripe\Control\Middleware\TrustedProxyMiddleware->process(SilverStripe\Control\HTTPRequest, Closure)
    HTTPMiddlewareAware.php:62
    SilverStripe\Control\Director->SilverStripe\Control\Middleware\{closure}(SilverStripe\Control\HTTPRequest)
    HTTPMiddlewareAware.php:65
    SilverStripe\Control\Director->callMiddleware(SilverStripe\Control\HTTPRequest, Closure)
    Director.php:382
    SilverStripe\Control\Director->handleRequest(SilverStripe\Control\HTTPRequest)
    HTTPApplication.php:114
    SilverStripe\Control\HTTPApplication::SilverStripe\Control\{closure}(SilverStripe\Control\HTTPRequest)
    call_user_func(Closure, SilverStripe\Control\HTTPRequest)
    HTTPApplication.php:137
    SilverStripe\Control\HTTPApplication->SilverStripe\Control\{closure}(SilverStripe\Control\HTTPRequest)
    HTTPMiddlewareAware.php:65
    SilverStripe\Control\HTTPApplication->callMiddleware(SilverStripe\Control\HTTPRequest, Closure)
    HTTPApplication.php:130
    SilverStripe\Control\HTTPApplication->execute(SilverStripe\Control\HTTPRequest, Closure, )
    HTTPApplication.php:113
    SilverStripe\Control\HTTPApplication->handle(SilverStripe\Control\HTTPRequest)
    index.php:24

```

GuySartorelli commented:
Huh. For some reason `DB::connection_attempted()` was returning false even though we had a connection.
Fixed now by removing that call. The call to `DB::is_active()` will still prevent us from asking the db if there's a table before we've made a db connection.

emteknetnz reviewed (changes requested):
I'm getting the following error on dev/build when I attempt to use memcached

Note there's a good chance one of my setup steps is incorrect. I've never used memcached before

```
In InjectionCreator.php line 20:
                                                          
  [SilverStripe\Core\Injector\InjectorNotFoundException]  
  Class `SS_SESSION_CACHE_FACTORY` does not exist         
```

```
# .env
SS_SESSION_SAVE_HANDLER_CLASS="SilverStripe\Control\SessionHandler\CacheSessionHandler"
SS_SESSION_CACHE_FACTORY="SilverStripe\Core\Cache\MemcachedCacheFactory"
SS_MEMCACHED_DSN="memcached://localhost:11211"
```

I setup memcached in my docker container as follows, as ssh'd in as root user
```
apt update && apt install -y memcached php8.3-memcached
memcached -d -m 64 -p 11211 -u memcache
```

```
Exception trace:
  at /var/www/vendor/silverstripe/framework/src/Core/Injector/InjectionCreator.php:20
 SilverStripe\Core\Injector\InjectionCreator->create() at /var/www/vendor/silverstripe/framework/src/Core/Injector/Injector.php:631
 SilverStripe\Core\Injector\Injector->instantiate() at /var/www/vendor/silverstripe/framework/src/Core/Injector/Injector.php:1027
 SilverStripe\Core\Injector\Injector->getNamedService() at /var/www/vendor/silverstripe/framework/src/Core/Injector/Injector.php:979
 SilverStripe\Core\Injector\Injector->get() at /var/www/vendor/silverstripe/framework/src/Core/Injector/Injector.php:623
 SilverStripe\Core\Injector\Injector->instantiate() at /var/www/vendor/silverstripe/framework/src/Core/Injector/Injector.php:1027
 SilverStripe\Core\Injector\Injector->getNamedService() at /var/www/vendor/silverstripe/framework/src/Core/Injector/Injector.php:979
 SilverStripe\Core\Injector\Injector->get() at /var/www/vendor/silverstripe/framework/src/Core/Injector/Injector.php:521
 SilverStripe\Core\Injector\Injector->convertServiceProperty() at /var/www/vendor/silverstripe/framework/src/Core/Injector/Injector.php:513
 SilverStripe\Core\Injector\Injector->convertServiceProperty() at /var/www/vendor/silverstripe/framework/src/Core/Injector/Injector.php:497
 SilverStripe\Core\Injector\Injector->updateSpecConstructor() at /var/www/vendor/silverstripe/framework/src/Core/Injector/Injector.php:1023
 SilverStripe\Core\Injector\Injector->getNamedService() at /var/www/vendor/silverstripe/framework/src/Core/Injector/Injector.php:979
 SilverStripe\Core\Injector\Injector->get() at /var/www/vendor/silverstripe/framework/src/Control/Session.php:209
 SilverStripe\Control\Session::getSaveHandler() at /var/www/vendor/silverstripe/framework/src/Control/SessionHandler/DbBuildSessionExtension.php:56
 SilverStripe\Control\SessionHandler\DbBuildSessionExtension->getSessionHandler() at /var/www/vendor/silverstripe/framework/src/Control/SessionHandler/DbBuildSessionExtension.php:42
 SilverStripe\Control\SessionHandler\DbBuildSessionExtension->onAfterBuild() at /var/www/vendor/silverstripe/framework/src/Core/Extension.php:141
 SilverStripe\Core\Extension->invokeExtension() at /var/www/vendor/silverstripe/framework/src/Core/Extensible.php:437
 SilverStripe\Dev\Command\DbBuild->extend() at /var/www/vendor/silverstripe/framework/src/Dev/Command/DbBuild.php:178
 SilverStripe\Dev\Command\DbBuild->doBuild() at /var/www/vendor/silverstripe/framework/src/Dev/Command/DbBuild.php:78
 SilverStripe\Dev\Command\DbBuild->execute() at /var/www/vendor/silverstripe/framework/src/Dev/Command/DevCommand.php:37
 SilverStripe\Dev\Command\DevCommand->run() at /var/www/vendor/silverstripe/framework/src/Cli/Command/PolyCommandCliWrapper.php:39
 SilverStripe\Cli\Command\PolyCommandCliWrapper->execute() at /var/www/vendor/symfony/console/Command/Command.php:318
 Symfony\Component\Console\Command\Command->run() at /var/www/vendor/symfony/console/Application.php:1092
 Symfony\Component\Console\Application->doRunCommand() at /var/www/vendor/silverstripe/framework/src/Cli/Sake.php:219
 SilverStripe\Cli\Sake->doRunCommand() at /var/www/vendor/symfony/console/Application.php:341
 Symfony\Component\Console\Application->doRun() at /var/www/vendor/symfony/console/Application.php:192
 Symfony\Component\Console\Application->run() at /var/www/vendor/silverstripe/framework/src/Cli/Sake.php:118
 SilverStripe\Cli\Sake->run() at /var/www/vendor/silverstripe/framework/bin/sake:19
 include() at /var/www/vendor/bin/sake:119

db:build [--no-populate] [--dont_populate]
```

GuySartorelli commented:
I think I must have only tested that before setting up the env var. The new injector config doesn't seem to be working based on that first error. I'll look into it.

GuySartorelli committed:
ENH Allow backticks in "factory" injector config

This allows us to use environment variables to set a factory class name
in injector configuration.

GuySartorelli commented:
For some reason with the cached session handler it's forcing dev/build if I'm not authenticated (after fixing the above).... will look into that now.

GuySartorelli commented:
Rebased in case something where's happening as a result of recent changes

GuySartorelli commented:
The dev/build thing is caused by https://github.com/silverstripe/silverstripe-framework/issues/11806 - I'll need to rebase again after the fix is merged.
